### ISSUE TITLE
Linux  runtime only uses 1 core of those available when running with forced scheduler affinity

### ISSUE THREAD

**Issue by RobertHenry6bev** [dotnet/runtime#81289]
### Description

I boot Linux using the isolcpus=I parameter, where I is a set of CPU cores that Linux is not otherwise to use for scheduling work on.

I run my dotnet program using the taskset wrappr: /usr/bin/taskset --cpu-list=J ...dotnetbinary...  where J is the set of CPU cores to run the binary on; by design of CPU affinity in Linux, then the affinitized binary can run on any of the cores, including those otherwises isolated by the isolcpus= kernel boot parameter.

If I & J is not empty (eg, there's a set intersection), the dotnet runtime only appears to run worker threads on the lowest numbered core, and the work is never parallelized across the set J.  I do see NET Server GC threads running on the other cores.

If I & J is empty, then the dotnet runtime happily inflates itself across all cores in set J, and the app runs in parallel.

Here's the trimmed output of ps showing the core of each dotnet thread in my workload.  The set I (isolated cores) was {1..50}, and the set J (asked-for cores) was {30, 31, 32, 33}.

The whole point of booting with isolcpus= is to control the amount of background work on the cores running (and measuring) the desired workload.
```
30 PlatformBenchma  PlatformBenchmarks
30 PlatformBenchma  PlatformBenchmarks
30 PlatformBenchma  PlatformBenchmarks
30 PlatformBenchma  PlatformBenchmarks
30 PlatformBenchma  PlatformBenchmarks
30 .NET Server GC   PlatformBenchmarks
30 .NET Finalizer   PlatformBenchmarks
30 PlatformBenchma  PlatformBenchmarks
30 .NET Timer       PlatformBenchmarks
30 .NET ThreadPool  PlatformBenchmarks
30 .NET ThreadPool  PlatformBenchmarks
30 Kestrel Timer    PlatformBenchmarks
30 .NET Sockets     PlatformBenchmarks
30 .NET ThreadPool  PlatformBenchmarks
30 .NET ThreadPool  PlatformBenchmarks
30 .NET ThreadPool  PlatformBenchmarks
30 .NET ThreadPool  PlatformBenchmarks
30 .NET ThreadPool  PlatformBenchmarks
30 .NET BGC         PlatformBenchmarks
30 .NET BGC         PlatformBenchmarks
30 .NET BGC         PlatformBenchmarks
30 .NET BGC         PlatformBenchmarks
31 .NET Server GC   PlatformBenchmarks
32 .NET Server GC   PlatformBenchmarks
33 .NET Server GC   PlatformBenchmarks
```

### Reproduction Steps

1. Run on a machine with >= 40 cores, or else change core enumerations below.
2. Edit /etc/default/grub(and possibly elsewhere) to add to GRUB_CMDLINE_LINUX_DEFAULT= the substring isolcpus=30,31,32,33
3. run `sudo update-grub`
4. reboot
5. cat /proc/cmdline to double check
6. examine output of `ps avx` to make sure that only processes named with [...] are running on  the isolcpus
7. /usr/bin/taskset --all-tasks --cpu-list=34,35,36,37 $workload
8. ps -A -L --format cpuid,lwp,pid,ppid,pcpu,cpuid,comm,nice,command | grep workload-process-name
9. observe many threads running on all cores
10. /usr/bin/taskset --all-tasks--cpu-llist=33,34,35,36
11. observe worker threads only on one core, probably core 33 (which is in the isolcpus set)
12. /usr/bin/taskset --all-tasks --cpu-list=30,31,32,33
13. observe only one thread doing work

### Expected behavior

All cores in the affinity set given by taskset are running and doing work; top shows %CPU for the workload at 400%, eg, not stuck a 100%.

### Actual behavior

workload only gets 100% of a cpu, not 400% (reflecting the 4 cores taskset was told to use)

### Known Workarounds

none

### Configuration

dotnet 7.0.2
Ubuntu 22.04.1 LTS
linux kernel 5.15.0-58-generic
ARM64 host machine  (I have not tried running on amd64/x86_64)

**Comment by AndyAyersMS** [comment]
@kouvel any chance you could look into this one?

**Comment by kouvel** [comment]
Discussed offline. I was able to repro the issue and also saw slightly different behavior at times, the issue seems to be with where threads are being scheduled rather than with .NET, confirmed similar behavior with a small C++ app. Based on [this](https://www.kernel.org/doc/html/latest/admin-guide/kernel-parameters.html), the `isolcpus` parameter is deprecated. Closing for now, as this doesn't seem to be a .NET issue.
